retry_delay.go

```diff
@@ -26,6 +27,7 @@ type retryDelayStrategy struct {
 	resetInterval time.Duration
 	retryCount    int
 	goodSince     time.Time // nonzero only if the state is currently "good"
+	lock          sync.Mutex
```

Comment:
Can we have explicit documentation about what members this lock protects? Would it be "all members" for now?

Reply:
It would, by convention, refer to anything that could ever be written by one goroutine and read by another. Anything that is read-only after a thing's been created, or that can only ever be accessed by one goroutine, wouldn't need to use the lock, but there would also be no harm in using the lock so that should be the default assumption. In other words, I think the case where you would want explicit documentation would be "here is why this field _doesn't_ need the lock". Adding a comment like "this lock is used for X and Y" is dangerous because it's extremely easy to add other things elsewhere and forget to update that comment.